3 - PB/Documentazione esterna/Specifica Tecnica/Specifica Tecnica.typ

```diff
@@ -146,7 +147,199 @@ Ciò avviene prestando attenzione alle tecnologie impiegate sia nel front-end ch
 
-== Diagramma delle classi
+== Struttura
+Per garantire modularità e separazione delle responsabilità, il programma adotta l'architettura "layered".
+Nello specifico vengono individuati tre strati:
+
+- *Persistence layer*: gestisce l'accesso al database e fornisce gli stumenti dedicati alla lettura dei suoi dati persistenti. I dati letti vengono processati al fine di poter creare gli elementi del Business layer;
+
+- *Business layer*: si occupa di elaborare i dati ricevuti dallo strato di persistenza e applicare le regole di business definite. È responsabile di implementare la logica dell'applicazione in modo indipendente dalle tecnologie di persistenza e di presentazione utilizzate;
+
+- *Presentation layer*: permette di trasformare i dati elaborati dal Business layer e le informazioni in una forma comprensibile e accessibile agli utenti finali. Questo include la creazione di interfacce utente grafiche e visualizzazioni 3D degli elementi di interesse.
+
+
+== Design pattern utilizzati
+I Design pattern introdotti nell'architettura dell'applicazione sono:
+
+- *pattern Data mapper*: viene utilizzato per interpretare i dati letti del database mantenendo separate la logica di business dal layer di persistenza. Le classi relative a questo pattern fungono da intermediari tra l'applicazione e la sorgente dati e sono responsabili della conversione delle strutture dati atte alla persistenza in oggetti del dominio dell'applicazione;
+
+- *pattern Repository*: come per il precedente pattern, viene implementato per separare la logica di business dalla logica di accesso ai dati. Le classi relative a questo pattern eseguono operazioni di lettura, aumentando l'astrazione dei dettagli specifici della persistenza dei dati e permettendo all'applicazione di interagirvi in modo indipendente dal tipo di archivio sottostante;
+
+- *pattern Provider*: applicato nel contesto tecnologico del progetto, soprattutto rispetto all'utilizzo di React, vengono sfruttate delle API per permette di gestire e trasferire i dati attraverso l'albero dei componenti in modo strutturato, evitando il "prop drilling", ovvero l'effetto che si verifica nei casi in cui è necessario trasportare i dati attraverso più livelli di componenti, anche se alcuni di essi non ne necessitano;
+
+- *pattern Strategy*: consente di definire una famiglia di algoritmi, incapsularli in classi separate e renderli intercambiabili. In questo modo è possibile applicare l'algoritmo appropriato senza dover conoscere i dettagli implementativi.
+
+
+
+== Classi e Componenti
+Ciascun layer possiede il suo indipendente sistema di classi e componenti e prevede metodi per comunicare con i layer adiacenti.
+
+=== Persistence layer
+Mediante delle chiamate API vengono letti i dati utili all'applicazione da un database esterno e vengono forniti degli adeguati file Json.
+Questi ultimi vengono utilizzati dalle classi derivate dall'adozione del pattern "Data mapper":
+- l'interfaccia DataMapperInterface;
```

Comment:
Non vorrei sbagliarmi ma i nomi di metodi, classi, parametri devono essere messi in `Monospace`